mithril-core/src/merkle_tree.rs

```diff
@@ -160,7 +247,122 @@ impl<D: Clone + Digest> MerkleTreeCommitment<D> {
     }
 }
 
-impl<D: Digest> MerkleTree<D> {
+impl<D: Clone + Digest> MerkleTreeCommitmentBatchCompat<D> {
+    /// Serializes the Merkle Tree commitment together with a message in a single vector of bytes.
+    /// Outputs `msg || self` as a vector of bytes.
+    // todo: Do we need to concat msg to whole commitment (nr_leaves and root) or just the root?
+    pub fn concat_with_msg(&self, msg: &[u8]) -> Vec<u8>
+    where
+        D: Digest,
+    {
+        let mut msgp = msg.to_vec();
+        let mut bytes = self.root.clone();
+        msgp.append(&mut bytes);
+
+        msgp
+    }
+
+    /// Check a proof of a batched opening. The indices must be ordered.
+    ///
+    /// # Error
+    /// Returns an error if the proof is invalid.
+    // todo: Update doc.
+    // todo: Simplify the algorithm.
```

Comment:
Let's add for future us
```suggestion
    // todo: Simplify the algorithm.
    // todo: Maybe we want more granular errors, rather than only `BatchPathInvalid`
```